flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala

```diff
@@ -19,45 +19,56 @@
 package org.apache.flink.table.planner.plan.schema
 
 import org.apache.flink.table.catalog.CatalogTable
-import org.apache.flink.table.planner.calcite.FlinkTypeFactory
 import org.apache.flink.table.planner.plan.stats.FlinkStatistic
-import org.apache.flink.table.planner.sources.TableSourceUtil
 import org.apache.flink.table.sources.{TableSource, TableSourceValidation}
-import org.apache.calcite.rel.`type`.{RelDataType, RelDataTypeFactory}
+
+import org.apache.calcite.rel.`type`.RelDataType
 import org.apache.flink.shaded.guava18.com.google.common.base.Preconditions
 import org.apache.flink.table.api.{TableException, WatermarkSpec}
-import org.apache.flink.table.types.logical.{TimestampKind, TimestampType}
+import org.apache.flink.table.planner.calcite.FlinkToRelContext
+
+import org.apache.calcite.plan.{RelOptSchema, RelOptTable}
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.logical.LogicalTableScan
+
+import java.util.{List => JList}
 
 import scala.collection.JavaConverters._
+import scala.collection.JavaConversions._
 
 /**
-  * Abstract class which define the interfaces required to convert a [[TableSource]] to
-  * a Calcite Table
+  * Abstract class which define the implementations required to translate
+  * the Calcite [[RelOptTable]] to the Flink specific relational expression with [[TableSource]],
+  * i.e. The computed column and watermark strategy specifications.
+  *
+  * <p>It also defines the [[copy]] method used for push down rules.
   *
-  * @param tableSource The [[TableSource]] for which is converted to a Calcite Table.
-  * @param isStreamingMode A flag that tells if the current table is in stream mode.
-  * @param statistic The table statistics.
+  * @param tableSource The [[TableSource]] for which is converted to a Calcite Table
+  * @param isStreamingMode A flag that tells if the current table is in stream mode
+  * @param catalogTable Catalog table where this table source table comes from
   */
 class TableSourceTable[T](
+    relOptSchema: RelOptSchema,
+    names: JList[String],
+    rowType: RelDataType,
```

Comment:
-1 to hard code `rowType` out of `TableSourceTable`. This is error-prone.

It's hard to make sure the the user-defined rowtype is the same as the output type of `TableSourceTable`,
unless user use the right util method and the right parameters to compute the rowtype, e.g. there are 3 `getSourceRowType` in `TableSourceUtil`, and use `CatalogTable#tableSchema` or `TableSource#tableSchema` as the parameter will get different/wrong rowType.

The `rowType` of `TableSourceTable` can be and should be self-defined. The constructor of `TableSourceTable` shouldn't care about the rowType, because all of the constructor parameters will affect the `rowType`, incl. `tableSource`, and `tableSource` might change (see `copy(tableSource: TableSource[_], statistic: FlinkStatistic)`).

Reply:
There is no any different, the `getRowType` logic does not change compared to before, and there should not be multiple implementation, if there is, it is error prone.

Reply:
But there is,  there are at least 2 different construction of `rowType`.
The [first one](https://github.com/danny0405/flink/blob/rr/flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.scala#L104) in `PushProjectIntoTableSourceScanRule` which gets `TableSchema` from `TableSource`. And the [second one](https://github.com/danny0405/flink/blob/rr/flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/CatalogSchemaTable.java#L183) in `CatalogSchemaTable` which gets `TableSchema` from `CatalogBaseTable`

Reply:
Agree, update to use `PushProjectIntoTableSourceScanRule` to use `tableSourceTable.catalogTable.getSchema`

Reply:
Using `tableSourceTable.catalogTable.getSchema` is wrong here...  This is the reason why we have FLINK-14716. 

My point is, this is an example where user might pass in wrong rowType, as even you may be confused on this.

Reply:
Finally we add a new method TableSourceTable#copy(.. selectedFields: Array[Int]) for the projected row type.